### DISCUSSION TITLE
[bug / unclear specs] Strings starting with '1' (0x31) are identical to strings starting with '𑁒' (\U00011052)

### DISCUSSION THREAD

**Discussion by Unknown6656** [dotnet/csharplang#7876] [General] [answered]
### Context
I am developing a small library, in which custom string to double parsing is employed. The parsing function also takes care of "unconventional" unicode symbols, i.e. numbers prefixed with `\u02D6`/`\uFE62`/`\uFF0B`/`\u208A`/`\u207A`  instead of `+`, etc.

### Behaviour
I noticed the following (possible) bug when filtering for unicode minus/hyphen signs:
```csharp
string one = 1.ToString();
bool starts = one.StartsWith("\U00011052"); // 𑁒
bool equals_inv = one.Equals("\U00011052", System.StringComparison.InvariantCulture);
bool equals = one.Equals("\U00011052");

System.Console.WriteLine(starts);
System.Console.WriteLine(equals_inv);
System.Console.WriteLine(equals);
```
prints the following output:
```
True
True
False
```
([sharplab](https://sharplab.io/#v2:C4LgTgrgdgPgAgRgAwAID2UCmKC8KEB0AKmgMrBgCWUA5gBQCUA3ALABQARmmgDYoDOwAIZhg/XOiwFyIsQHVKwABZ0ARAB0Aqkh0JkAVgBMq5igD0ZlIBANwEo77LrxSYAjhCE9+AfWoA3CRkwCAFFXd341LR0kPSQjVQAaFERpCmoaAGE0AFsABxFKfgwCAEkoH3yhKGB0iB5gCDBMZntuPhc3D38pEI7wjW1dA2NmtnZkxABOOkFZfhHxhCn2sO8y+cJJumWPZiA=))

### Unclear Specs?
Now, I get that https://www.compart.com/en/unicode/U+11052 defines the codepoint as "Brahmi Number One", however, I'm not sure why '1' and '𑁒' are considered equal due to their different optical appearance. Furthermore, the documentation on `String.StartsWith(String)` does not document the varying behaviour in regards to `String.Equals(String)` when not supplied with an explicit value for `StringComparison`. (see https://learn.microsoft.com/en-us/dotnet/api/system.string.startswith?view=net-8.0#system-string-startswith(system-string))

### Question
Would it be possible to clarify this behaviour somehow in the C# docs? I mean either the comparison on non-latin number symbols or the explanation of the following behaviour:
```csharp
string A = "1";
string B = "𑁒";

_ = A.StartsWith(B); // True
_ = A.Equals(B); // False
```

**Comment by huoyaoyuan** [comment, marked as answer]
The behavior of string operations belongs to [dotnet/runtime](https://github.com/dotnet/runtime) repo. It's not a part of the language.

It is an inconsistency that `Equals` defaults to ordinal while `StartsWith` defaults to culture-aware. Unfortunately, this cannot be easily addressed due to the huge compatibility debt for 25 years. There was discussion about helping this at https://github.com/dotnet/runtime/issues/43956.

Meanwhile, you can use [globalization analyzers](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/globalization-warnings) to help you specify behavior explicitly, or simply check the source code at source.dot.net about the option-less overload. Documentation improvement requests should be sent at [dotnet/docs](https://github.com/dotnet/docs).

**Reply by Unknown6656** [reply]
Thanks, I submitted an issue to dotnet/runtime (https://github.com/dotnet/runtime/issues/97544) and I'll transfer it to dotnet/docs if needed.
